src/window/window_wrapper.rs

```diff
@@ -236,10 +236,44 @@ impl WinitWindowWrapper {
                 }
             }
             #[cfg(target_os = "macos")]
-            WindowSettingsChanged::Transparency(opacity) => {
-                log::info!("transparency changed to {}", opacity);
-                invalidate_shadow(self.windowed_context.window());
-                set_window_blurred(self.windowed_context.window(), opacity);
+            WindowSettingsChanged::BackgroundColor(background_color) => {
+                log::info!("background_color changed to {}", background_color);
+                let window_settings = SETTINGS.get::<WindowSettings>();
+                let transparency = window_settings.transparency;
+                let show_border = window_settings.show_border;
+                self.macos_feature.set_background(
+                    transparency,
+                    show_border,
+                    background_color,
+                    false,
+                );
+            }
+            #[cfg(target_os = "macos")]
+            WindowSettingsChanged::ShowBorder(show_border) => {
+                log::info!("show_border changed to {}", show_border);
+                let window_settings = SETTINGS.get::<WindowSettings>();
+                let transparency = window_settings.transparency;
+                let background_color = window_settings.background_color;
+                self.macos_feature.set_background(
+                    transparency,
+                    show_border,
+                    background_color,
+                    true,
+                );
+            }
+            #[cfg(target_os = "macos")]
+            WindowSettingsChanged::Transparency(transparency) => {
+                log::info!("transparency changed to {}", transparency);
+                let window_settings = SETTINGS.get::<WindowSettings>();
+                let show_border = window_settings.show_border;
+                let background_color = window_settings.background_color;
+                self.macos_feature.set_background(
+                    transparency,
+                    show_border,
+                    background_color,
+                    true,
+                );
+                set_window_blurred(self.windowed_context.window(), transparency);
             }
```

Comment:
Maybe we should add `macos_feature.handle_windows_settings` and move all these there (and also the WindowBlurred function?

I think that would be cleaner than adding `#[cfg(target_os = "macos")]` to each match.

And perhaps also read the window settings inside the functions themselves, the match is easier to read if it's kept short.

Reply:
I particularly finds the older version cleaner to read, even if it has little more lines.

Isolated behavior for each setting, easier to debug, `log.info` has been removed... or maybe I am wrong.

It feels like overengineering our code trying to be as more generic as possible where we don't need to? or is just not the right moment?

A mix of feelings.

Reply:
Yeah, matching `changed_setting` twice (once for all platforms and once for macOS) isn't super pretty. But it has two advantages:
1. macOS specific config updates are now all handled in `macos.rs`
2. shared logic for all platforms while still allowing additional platform specific updates/overrides

About 1.: I initially loved the idea of having all macOS specific window settings inside `macos.rs` but thinking about it again there is something nice about not having to jump between files and functions to read the code for what is arguably the same concern (i.e. observing window setting changes). It probably isn't more idiomatic than just doing `#[cfg(target_os = "macos")]`. Reverted this in 22b88b3.

And yes, I did remove `log.info` (oops) as it all went through the same branch now:
```rust
match changed_setting {
    WindowSettingsChanged::BackgroundColor(_)
    | WindowSettingsChanged::ShowBorder(_)
    | WindowSettingsChanged::Transparency(_)
    | WindowSettingsChanged::WindowBlurred(_) => {
        self.update_background(window, ignore_deprecation_warning);
    }
    _ => {}
}
```

Reply:
usually I use this kind of multiple match pattern for testing purposes or decoding many similar range of types that has similar ends. 

for our use case I much prefer the reverted code you had just done.

LGTM as it is now.

Reply:
The issue with mixing code from different platforms is more a maintainability issue, than a readability one here.

Most people don't have the ability to compile and test on macOS, so it's easy to refactor some code and forget to change the macOS parts. That has happened both to me and others before, so it's a real problem. Now, isolating the platform parts into their own files don't make us immune to the problem, but the breakages get more limited, and apply to only the API that the platform specific code use. But you are more free to refactor the rest of the code, split functions, rename local variables and so on.

It also causes problems with warnings for unused variables, unused use declarations and so on. 

That said, this latest iteration is far better than the original code, mostly due to moving the blur code away to where it belongs.  But also, by removing the noise from the match statements itself. So, I can accept https://github.com/neovide/neovide/pull/2351/commits/22b88b3e7a5c2027d715d849d38ac19e413033c0, but I personally prefer the version before that, also for the reason that the Window wrapper itself have no reason to deal with those events at all, they are completely unrelated to the functionality of it.

But it can become a readability issue as well, here it only works because, we are only special casing macOS here, if more platforms need special cases, especially for the same settings, it will quickly become a mess. Look at https://github.com/neovide/neovide/blob/main/src/renderer/vsync/mod.rs, I have written that myself, but it's a mess. This one from https://github.com/neovide/neovide/pull/2359, is very slightly better, but only works because the code is really simple https://github.com/neovide/neovide/blob/8c2e80c9ff653bc56224a29d14bfc6a854ef7afe/src/renderer/vsync/mod.rs

Reply:
why we just don't avoid the multiple match pattern and just move it as it is to `macOS.rs` then?

I agree about the mess that it can become using only `#[cfg]`, definitely.

maybe this would be better for everyone.

and jesus @fredizzimo that makes my eyes hurt. I guess your color scheme should handle it as ghost texted.

![CleanShot 2024-02-23 at 01 01 24@2x](https://github.com/neovide/neovide/assets/33763843/b07bc9f6-9179-429e-b2df-4c3e9c1fa397)

Reply:
To put my two cents in, I think we should start a separate issue on this.

I briefly looked into it but this looks to be a large scale refactor.
For macos alone migrating to a separate module is big, due to the hierarchy between window creation, the window wrapper, the main loop, the macos window wrapper and how all of them currently use `#[cfg]`.

It might be worth to look at how [winit](https://github.com/rust-windowing/winit/blob/master/src/platform_impl/mod.rs) and [gpui](https://github.com/zed-industries/zed/tree/main/crates/gpui) do this.

For now, what is the consensus here? Move back to `macos.rs` or leave as-is?

Reply:
in case of moving to macos.rs I would avoid the multiple match pattern to keep the log.info for each setting changes for example and to split responsibilities of each event once this is interactive.

besides that, I have nothing else to add/share.

Reply:
A separate call is fine
`macos_window_feature.handle_settings_changed(changed_setting)`

That way the macos code csn independently decide which settings to handle.

Reply:
And if you rename it to something more generic, we could use the same pattern as winit, and it would become for example `platform_window.handle_settings_changed`